kubeadm: add a new output/v1alpha2 API; deprecate output/v1alpha1

neolit123 opened the pull request:
#### What this PR does / why we need it:

The output/v1alpha2 API is a copy of output/v1alpha1 with a minor difference
where output/v1alpha2.BootstrapToken embeds
bootstraptoken/v1.BootstrapToken instead of
kubeadm/v1beta2.BootstrapToken.

Embedding the later is an undesired binding between the "kubeadm"
and "output" groups, preventing the eventual deprecation and removal
of the kubeadm/v1beta2 API.

This new output API version, unlike v1alpha1, does not include
defaulting which is not needed.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/2427

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: add a new output/v1alpha2 API that is identical to the output/v1alpha1, but attempts to resolve some internal dependencies with the kubeadm/v1beta2 API. The output/v1alpha1 API is now deprecated and will be removed in a future release.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: add a new output/v1alpha2 API

The API is a copy of output/v1alpha1 with a minor difference
where output/v1alpha2.BootstrapToken embeds
bootstraptoken/v1.BootstrapToken instead of
kubeadm/v1beta2.BootstrapToken.

Embedding the later is an undesired binding between the "kubeadm"
and "output" groups, preventing the eventual deprecation and removal
of the kubeadm.v1beta2 API.

This new output API version, unlike v1alpha1, does not include
defaulting which is not needed.

neolit123 committed:
kubeadm: use output/v1alpha2 in kubeadm commands

Use the new API for "config images list" and
"token list".

neolit123 commented:
/kind cleanup feature deprecation
/priority important-longterm
/triage accepted

neolit123 commented:
/retest

neolit123 commented:
/milestone v1.23

neolit123 committed:
kubeadm: set output/v1alpha2 as the priority version

neolit123 committed:
kubeadm: mark output/v1alpha1 as deprecated

neolit123 committed:
kubeadm: make output/BootstrapToken use bootstraptoken/v1

neolit123 committed:
kubeadm: update converters after the output/v1alpha2 addition

The addition of output/v1alpha2 made the converter-gen require
an explicit converter for:
kubeadm/v1beta2.BootstrapToken -> bootstraptoken/v1.BootstrapToken.

Add this converter under kubeadm/v1beta.
Use the converter in output/v1alpha1.

neolit123 commented:
/hold for review

fabriziopandini commented on cmd/kubeadm/app/apis/kubeadm/v1beta2/conversion.go:
```diff
@@ -64,3 +67,14 @@ func Convert_v1beta2_JoinConfiguration_To_kubeadm_JoinConfiguration(in *JoinConf
 	return nil
 }
+
+// Convert_v1beta2_BootstrapToken_To_v1_BootstrapToken is required so that
```
Comment seems truncated

neolit123 replied:
indeed, now i need to remember what it was about.

neolit123 replied:
updated.

fabriziopandini reviewed (commented):
two small nits, otherwise lgtm

neolit123 commented on cmd/kubeadm/app/apis/output/v1alpha2/register.go:
```diff
@@ -0,0 +1,67 @@
+var SchemeGroupVersion = schema.GroupVersion{Group: GroupName, Version: "v1alpha2"}
+
+var (
+	// TODO: move SchemeBuilder with zz_generated.deepcopy.go to k8s.io/api.
+	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
```
i don't think so. changes there are hard to argue about, thus i can just remove this TODO.

neolit123 replied:
added separate commit `kubeadm: remove TODO about moving SchemeBuilder to k8s.io/api` for that.
(i prefer if we keep the commit set here instead of squashing)

neolit123 committed:
kubeadm: remove TODO about moving SchemeBuilder to k8s.io/api

This TODO is no longer in scope, thus remove it from
all register.go files under /app/apis.

fabriziopandini commented:
/lgtm
@neolit123 feel free to un hold

neolit123 commented:
/hold cancel
